### PR TITLE
C#: Fixes for AST printing

### PR DESCRIPTION
**Pull request by tamasvajk**

### CODE REVIEW

**Comment by tamasvajk** [comment]
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/519/

**Comment by tamasvajk** [comment]
> ~https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/519/~
> https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/524/

I'm surprised to see the changes in the diffs. None of them seem to be attribute and namespace declaration related though.

File: csharp/ql/test/experimental/ir/ir/PrintAst.expected [hunk 1]
```diff
@@ -5,7 +5,7 @@ array.cs:
 #    3|     4: [BlockStmt] {...}
 #    4|       0: [LocalVariableDeclStmt] ... ...;
 #    4|         0: [LocalVariableDeclAndInitExpr] Int32[] one_dim = ...
-#    4|           -1: [TypeMention] Int32[]
+#    4|           -1: [TypeMention] int[]
 #    4|             1: [TypeMention] int
```
Comment (tamasvajk) [inline csharp/ql/test/experimental/ir/ir/PrintAst.expected]: @hvitved The main motivation for this commit was to unify the type name printing of `TypeMention`s. I think this looks a bit better, but we're still showing `Int32[]` in the above `LocalVariableDeclAndInitExpr` (and in some other nodes too). Should/Can I change those classes to use the `Type.toStringWithTypes`? It feels a bit strange to change all these base class `toString()`s just for the sake of AST printing.

What was the initial idea behind `toString` vs `toStringWithTypes`?

File: csharp/ql/test/library-tests/types/PrintAst.expected [hunk 2]
```diff
@@ -50,29 +50,29 @@ types.cs:
...
-#   29|         1: [TypeMention] Nullable<Byte>
-#   29|           1: [TypeMention] Byte
+#   29|       -1: [TypeMention] Nullable<Byte>[]
+#   29|         1: [TypeMention] byte?
+#   29|           1: [TypeMention] byte
 #   30|     30: [Method] NullableArrayArrayType
-#   30|       -1: [TypeMention] Nullable[]
-#   30|         1: [TypeMention] Nullable<Byte>
-#   30|           1: [TypeMention] Byte
+#   30|       -1: [TypeMention] Nullable<Byte>[][]
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/types/PrintAst.expected]: `Nullable[]` vs `Nullable<Byte>[][]` is the reason why `toStringWithTypes()` was used.

**Comment by tamasvajk** [comment]
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/539/

File: csharp/extractor/Semmle.Extraction.CSharp/Analyser.cs [hunk 3]
```diff
@@ -230,7 +230,7 @@ private void DoAnalyseCompilation(string cwd, string[] args)
                 var transformedAssemblyPath = PathTransformer.Transform(assemblyPath);
                 var assembly = compilation.Assembly;
                 var projectLayout = layout.LookupProjectOrDefault(transformedAssemblyPath);
-                var trapWriter = projectLayout.CreateTrapWriter(Logger, transformedAssemblyPath, true, options.TrapCompression);
+                var trapWriter = projectLayout.CreateTrapWriter(Logger, transformedAssemblyPath, options.TrapCompression, discardDuplicates: true);
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Analyser.cs:233]: Wasn't this where we agreed it should be `discardDuplicates: false`?

Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Analyser.cs:233]: Ah, you're right. I messed up the rebase. Thanks for finding this.

**Comment by tamasvajk** [comment]
The differences job revealed a couple of type mention extraction failures...

File: csharp/ql/test/library-tests/csharp8/PrintAst.expected [hunk 4]
```diff
@@ -491,9 +461,6 @@ NullableRefTypes.cs:
 #  131|   16: [Field] l
 #  131|     -1: [TypeMention] MyClass[,,][,,,][][,]
 #  131|       1: [TypeMention] MyClass
-#  131|         1: [TypeMention] MyClass
-#  131|           1: [TypeMention] MyClass
-#  131|             1: [TypeMention] MyClass
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/csharp8/PrintAst.expected]: This corresponds to `MyClass?[,,,][][,]?[,,] l;`.

The type mention hierarchy is now similar to `MyClass[,,,][][,][,,] k;` a couple of lines above, that is, there's only one type mention for the multi dimensional array and one for the element type.

**Comment by tamasvajk** [comment]
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/543/

**Comment by tamasvajk** [comment]
Note to self: add change notes.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/TypeMention.cs [hunk 5]
```diff
@@ -23,27 +23,36 @@ private TypeMention(Context cx, TypeSyntax syntax, IEntity parent, Type type, Mi
             this.loc = loc;
         }
 
-        private static TypeSyntax GetElementType(TypeSyntax type)
+        private TypeSyntax GetArrayElementType(TypeSyntax type)
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/TypeMention.cs:26]: Perhaps `GetUnderlyingType` is a better name?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/TypeMention.cs [hunk 6]
```diff
@@ -54,19 +63,27 @@ protected override void Populate(TextWriter trapFile)
...
+                        }
+                        else
+                        {
+                            Create(cx, nts.ElementType, parent, type);
+                        }
                     }
                     else if (type is ArrayType array)
                     {
-                        Create(cx, nts.ElementType, parent, array);
+                        Create(cx, nts.ElementType, parent, type);
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/TypeMention.cs:86]: Why this change?

File: csharp/ql/test/library-tests/definitions/definitions.expected [hunk 7]
```diff
@@ -53,10 +53,10 @@
 | definitions.cs:114:28:114:28 | T | definitions.cs:114:17:114:17 | T | T |
 | definitions.cs:114:32:114:32 | A | definitions.cs:97:11:97:11 | A | T |
 | definitions.cs:117:27:117:27 | T | definitions.cs:117:18:117:18 | T | T |
-| definitions.cs:119:20:119:21 | I2<Object> | definitions.cs:117:15:117:19 | I2<> | T |
+| definitions.cs:119:20:119:29 | I2<Object> | definitions.cs:117:15:117:19 | I2<> | T |
 | definitions.cs:121:18:121:18 | A | definitions.cs:97:11:97:11 | A | T |
 | definitions.cs:121:21:121:22 | I1 | definitions.cs:112:15:112:16 | I1 | T |
-| definitions.cs:121:25:121:26 | I2<A> | definitions.cs:117:15:117:19 | I2<> | T |
+| definitions.cs:121:25:121:29 | I2<A> | definitions.cs:117:15:117:19 | I2<> | T |
```
Comment (hvitved) [inline csharp/ql/test/library-tests/definitions/definitions.expected:59]: This change is problematic, as we will now have overlapping location spans for the character `A`: one for `I2<A>` (which jumps to `I2<>`) and one for `A` (which jumps to `A`), so not sure if the IDE/LGTM will work correctly when trying to jump to the definition for `A`. I agree that the right location for `I2<A>` is the whole span, so perhaps the fix instead is to adjust `hasLocationInfo` for constructed types to exclude type arguments, just like we [remove qualifiers from method calls](https://github.com/github/codeql/blob/main/csharp/ql/src/definitions.qll#L65).

File: csharp/ql/src/semmle/code/csharp/PrintAst.qll [hunk 8]
```diff
@@ -120,6 +120,16 @@ private predicate isInterestingBaseType(ValueOrRefType base) {
...
+  TTopLevelTypeMentionNode(TypeMention typeMention) {
+    shouldPrint(typeMention.getTarget(), _) and
+    not isNotNeeded(typeMention.getTarget().getParent*()) and
+    not typeMention.getTarget().getParent*() instanceof TypeParameter
+  } or
+  TNestedTypeMentionNode(TypeMention typeMention) {
+    shouldPrint(typeMention.getParent+().getTarget(), _) and
+    not isNotNeeded(typeMention.getParent+().getTarget().getParent*()) and
+    not typeMention.getParent+().getTarget().getParent*() instanceof TypeParameter
+  } or
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/PrintAst.qll:132]: Is there a good reason to distinguish nested and top-level type mentions? It seems that something like this would encompass both cases:
```ql
TTypeMentionNode(TypeMention typeMention) {
  exists(Element target |
    target = typeMention.getParent*().getTarget() and
    shouldPrint(target, _) and
    not isNotNeeded(target.getParent*()) and
    not target.getParent*() instanceof TypeParameter
  )
}
```

File: csharp/ql/src/semmle/code/csharp/PrintAst.qll [hunk 9]
```diff
@@ -238,6 +243,48 @@ abstract class ElementNode extends PrintAstNode, TElementNode {
...
+
+  /**
+   * Gets the `TypeMention` represented by this node.
+   */
+  final TypeMention getTypeMention() { result = typeMention }
+
+  override NestedTypeMentionNode getChild(int childIndex) {
+    result.getTypeMention() =
+      rank[childIndex](TypeMention t, Location l |
+        t = any(TypeMention tm | tm.getParent() = typeMention) and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/PrintAst.qll:266]: t.getParent() = typeMention

File: csharp/ql/src/semmle/code/csharp/PrintAst.qll [hunk 10]
```diff
@@ -263,8 +310,15 @@ class ControlFlowElementNode extends ElementNode {
...
 
-  override ElementNode getChild(int childIndex) {
-    result.getElement() = controlFlowElement.getChild(childIndex)
+  override PrintAstNode getChild(int childIndex) {
+    (
+      childIndex = min(int i | exists(controlFlowElement.getChild(i))) - 1
+      or
+      not exists(controlFlowElement.getAChild()) and childIndex = 0
+    ) and
+    result.(TopLevelTypeMentionNode).getTypeMention().getTarget() = controlFlowElement
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/PrintAst.qll:319]: It would probably make sense to add a `getTarget()` predicate to `TypeMentionNode` and call that instead (and more places further down).

File: csharp/ql/src/semmle/code/csharp/PrintAst.qll [hunk 11]
```diff
@@ -576,19 +690,21 @@ final class BaseTypesNode extends PrintAstNode, TBaseTypesNode {
...
-    result.getBaseType() =
-      rank[childIndex](ValueOrRefType base, string name |
-        base = valueOrRefType.getABaseInterface() and
-        name = base.toString()
+    result.getTypeMention() =
+      rank[childIndex](TypeMention t, Location l |
+        t =
+          any(TypeMention tm |
+            tm.getType() = valueOrRefType.getABaseInterface() and tm.getTarget() = valueOrRefType
+          ) and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/PrintAst.qll:703]: `t.getType() = valueOrRefType.getABaseInterface() and t.getTarget() = valueOrRefType`

**Review by hvitved — CHANGES_REQUESTED** [review]
A few comments, otherwise LGTM.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/NamespaceDeclaration.cs [hunk 12]
```diff
@@ -41,8 +46,24 @@ protected override void Populate(TextWriter trapFile)
...
         }
 
-        public static NamespaceDeclaration Create(Context cx, NamespaceDeclarationSyntax decl, NamespaceDeclaration parent) => new NamespaceDeclaration(cx, decl, parent);
+        public static NamespaceDeclaration Create(Context cx, NamespaceDeclarationSyntax decl, NamespaceDeclaration parent)
+        {
+            var init = (decl, parent);
+            return NamespaceDeclarationFactory.Instance.CreateEntity(cx, init, init);
+        }
+
+        class NamespaceDeclarationFactory : ICachedEntityFactory<(NamespaceDeclarationSyntax decl, NamespaceDeclaration parent), NamespaceDeclaration>
```
[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/NamespaceDeclaration.cs:55]: I think it's definitely enough to use the namespace as the cache key, so 
```
var init = (decl, parent);
return NamespaceDeclarationFactory.Instance.CreateEntity(cx, decl, init);
```
would work a couple of lines above.

At the same time, how would we populate the `parent_namespace_declaration` relation if the parent is not part of the `init`? Would we get the parent from the cache, like this? I'll give it a try.
```
if (node.Parent is NamespaceDeclarationSyntax p)
{
    var parent = NamespaceDeclaration.Create(Context, p);
    if (parent != null)
    {
        trapFile.parent_namespace_declaration(this, parent);
    }
}
```

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/TypeMention.cs (hunk 5 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/TypeMention.cs:26]: I thought about this too. This method should not be called from `case SyntaxKind.NullableType` in `Populate`. It should only be used to get the element type of array-like `TypeSyntax` elements (array, pointer), hence the name. However, it's recursive and during the recursion, we need to also handle `NullableTypeSyntax`, and this introduces a bit of confusion.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/TypeMention.cs (hunk 6 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/TypeMention.cs:86]: This is only cosmetics. This way this branch and the previous one are exactly the same. I'll change the `if` to `if (type is ArrayType)`

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/NamespaceDeclaration.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/NamespaceDeclaration.cs:55]: I still think we want to pass in the parent as `init`, we just don't need it for caching (i.e., exactly what you suggest in your first code snippet above).

File: csharp/ql/src/semmle/code/csharp/PrintAst.qll (hunk 8 above)
Comment (tamasvajk) [reply csharp/ql/src/semmle/code/csharp/PrintAst.qll:132]: We need to differentiate between the two, because we only want to have top level type mentions directly under `ElementNode`s and `BaseTypesNode`s.

Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/PrintAst.qll:132]: But that could be handled by restricting to those `TypeMentionNode`s for which `getTarget()` exists.

File: csharp/ql/test/library-tests/definitions/definitions.expected (hunk 7 above)
Comment (tamasvajk) [reply csharp/ql/test/library-tests/definitions/definitions.expected]: Thanks for noticing. I pushed a fix for this.

File: csharp/ql/src/semmle/code/csharp/PrintAst.qll (hunk 8 above)
Comment (tamasvajk) [reply csharp/ql/src/semmle/code/csharp/PrintAst.qll:132]: Makes sense. Let me adjust it.

**Review by hvitved — APPROVED** [review]
